Fix the 4.4.x build

gkalpak opened the pull request:
This PR backports (and adapts) relevant commits from #18722, #19511 and #19603 to 4.4.x.
It also builds on top of #19586 to fix the aio tests.
Finally it adds a couple of commits on top of that to fix `NgPackagesInstaller` testing and some 4.4.x-specific issues.

gkalpak committed:
build(aio): add support for using the locally built Angular packages for aio

This commit allows building angular.io against the locally built Angular
packages. It adds two new npm scripts:

- `setup-local`: Same as `setup`, but overwrites the Angular packages for both
  angular.io and the examples boilerplate with the locally built ones.
- `build-local`: Same as `build`, but uses `setup-local` instead of `setup`
  under the hood, thus overwriting installed Angular packages with locally built
  ones.

Fixes #18611

petebacondarwin committed:
fix(aio): fix SearchService to work with TypeScript 2.4

The call to `race` required a type parameter to disambiguate the return type.

petebacondarwin committed:
build(aio): upgrade `ts-node` to support newer TypeScript config

This was causing `Cannot find type definition file for 'jasmine'. (2688)`
errors when running Protractor.

petebacondarwin committed:
build(aio): ensure webdriver is updated when switching between local and npm deps

petebacondarwin committed:
docs(aio): fix animations example to work with Angular v5

petebacondarwin committed:
build(aio): ignore @angular/mobile-toolkit in dist builds

The new v5 version of this toolkit is too incompatible for us to run
AIO against both versions. So for now we will ignore this libary when
running against local Angular distributions and continue to use the
version on npm.

alxhub committed:
fix(aio): downgrade yarn to 1.0.2 temporarily

petebacondarwin committed:
ci(aio): use custom package.json to run with local distributables

Closes #19388

petebacondarwin committed:
build(aio): support ignoring dist packages in "local" mode

petebacondarwin committed:
build(aio): provide cleaner scripts for local build switching

petebacondarwin committed:
build(aio): example-boilerplate is no longer responsible for yarn install

The tooling for boilerplate was also running `yarn install` on the examples'
shared folder. But since this is handled by `ng-packages-installer` this
commit refactors the tools so that the boilerplate no longer does this
anymore.

gkalpak committed:
test(aio): fix testing of `NgPackagesInstaller`

mary-poppins commented:
You can preview 12ed639 at https://pr19600-12ed639.ngbuilds.io/.

mary-poppins commented:
You can preview bab8f3a at https://pr19600-bab8f3a.ngbuilds.io/.

gkalpak committed:
build(aio): do not fail on first `yarn setup`

gkalpak committed:
build(aio): fix overwriting with local Angular packages that depend on other local ones

mary-poppins commented:
You can preview 9ef671a at https://pr19600-9ef671a.ngbuilds.io/.

mary-poppins commented:
You can preview de17331 at https://pr19600-de17331.ngbuilds.io/.

mary-poppins commented:
You can preview b3e27d5 at https://pr19600-b3e27d5.ngbuilds.io/.

petebacondarwin commented on aio/package.json:
```diff
@@ -6,33 +6,45 @@
     "deploy-preview": "scripts/deploy-preview.sh",
     "deploy-production": "scripts/deploy-to-firebase.sh",
-    "check-env": "node scripts/check-environment",
+    "check-env": "yarn ~~check-env",
```
does this need a bit of `yarn install`?

gkalpak replied:
Yeah. I thought I applied #19603 on top of the other changes. Good catch :smiley:

petebacondarwin commented:
Needs some approval from the `build-and-ci` group.

petebacondarwin commented:
Travis e2e tests for AIO are not happy :-(

gkalpak commented:
It's the flakey upgrade tests 😞 They pass locally 😕 
I'll see what I can do...

mary-poppins commented:
You can preview 0e6cb49 at https://pr19600-0e6cb49.ngbuilds.io/.

mary-poppins commented:
You can preview cd3ef00 at https://pr19600-cd3ef00.ngbuilds.io/.

gkalpak committed:
test(aio): temporarily disable failing examples e2e tests

Temporarily disable a couple of failing upgrade example e2e tests
to unblock the 4.4.x branch while investigating and fixing the
issues.
It seems to be some typings-related issues that TypeScript 2.5.3
complains about.

(Note: The same tests are temporarily disabled on master too.)

gkalpak commented:
I ended up disabling the failing e2e tests (after I realized they are disabled on master too). It is some typings-related issue. (Opened #19625 to track it.)

mary-poppins commented:
You can preview a59f781 at https://pr19600-a59f781.ngbuilds.io/.

mary-poppins commented:
You can preview e88a6ec at https://pr19600-e88a6ec.ngbuilds.io/.

chuckjaz commented:
This PR has 16 commits and many don't look related to the fixing the PR. Can the commits that are covered by other PRs be removed from this?

gkalpak commented:
@chuck, not sure which commits you refer to. The only commit that is covered by another PR is the one I brought in from #19586. It will be a "no-op" if #19586 is merged first.

Other than that, this PR backports other PRs that have landed on master (which are necessary to correctly run the examples e2e tests against the local Angular build) and adds some fixes on top (some of them 4.4.x specific).

gkalpak commented:
This has been merged into 4.4.x.
